src/zkstar.rs

```diff
@@ -0,0 +1,118 @@
+use crate::errors::InternalError;
+use libpaillier::unknown_order::BigNumber;
+use serde::{Deserialize, Serialize};
+
+/// A warpper around the modulus of the group [ZStarN] is defined over
+/// This allows passing the modulus by reference and avoid serializing it
+#[derive(Serialize)]
+struct ZStarNBuilder {
+    N: BigNumber,
+}
+
+impl ZStarNBuilder {
+    pub fn new(n: BigNumber) -> Self {
+        ZStarNBuilder { N: n }
+    }
+    /// The only way to construct an element of [ZStarN] is to validate that [ZStarNUnverified]
+    /// is properly constructed with respect to the current [ZStarNBuilder]
+    pub fn validate(&self, unverified: ZStarNUnverified) -> Result<ZStarN, InternalError> {
+        if unverified.value().is_zero() {
+            return Err(InternalError::IsZero);
+        } else if unverified.value() > self.value() {
+            return Err(InternalError::LargerThanModulus);
+        }
+        let result = unverified.value().extended_gcd(self.value());
+        if result.gcd != BigNumber::one() {
+            return Err(InternalError::NotCoprime);
+        }
+        Ok(ZStarN {
+            value: unverified.value().clone(),
+            builder: self,
+        })
+    }
+    /// Return the modulus of [ZStarNBuilder]
+    fn value(&self) -> &BigNumber {
+        &self.N
+    }
+}
+
+/// An verifier element of [ZStarN]
+#[derive(Serialize)]
+struct ZStarN<'a> {
+    value: BigNumber,
+    builder: &'a ZStarNBuilder,
+}
+
+impl<'a> ZStarN<'a> {
+    /// Create new canonically-formed element in `ZStar`.
+    /// Constructor always verifies that the element is a member of ZStar
+    fn new(
+        unverified: ZStarNUnverified,
+        builder: &'a ZStarNBuilder,
+    ) -> Result<Self, InternalError> {
+        builder.validate(unverified)
+    }
+    /// Return the value stored in [ZStarN]
+    fn value(&self) -> &BigNumber {
+        &self.value
+    }
+    fn serialize(&self) -> Result<Vec<u8>, InternalError> {
+        serialize!(self.value())
+    }
+}
+
+/// User can only deserialize into a [ZStarNUnverified] instance. They must call tryfrom instance below to create a proper [ZStarN]
+#[derive(Deserialize)]
+struct ZStarNUnverified {
+    value: BigNumber,
+}
+impl ZStarNUnverified {
+    fn new(value: BigNumber) -> Self {
+        ZStarNUnverified { value }
+    }
+    fn value(&self) -> &BigNumber {
+        &self.value
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
question: Why are the tests in here called `_fast`? I usually lean toward names that are closer to documentation, since I usually read them in the terminal and want to quickly know what failed. e.g. `zkstar_constructor_rejects_elements_outside_group` or `zkstar_verification_works` for the first one.

Reply:
I was following some older naming conventions in `paillier.rs` where depending on whether or not you generate the primes safely or not you had a `_slow`, `_fast` suffix. I can definitely remove it.

Reply:
Ah, I see. We just recently updated the CI to run ignored tests after you merge to main, so any test that is both slow and redundant can be tagged with `#[ignore = "whatever reason"]`. So now we can assume that anything not ignored is reasonably fast!